sysdig/resource_sysdig_secure_cloud_auth_account.go

```diff
@@ -382,33 +405,106 @@ func constructAccountComponents(accountComponents []*cloudauth.AccountComponent,
 	accountComponents := constructAccountComponents([]*cloudauth.AccountComponent{}, data)
 
-	featureData := data.Get("feature").(interface{})
+	featureData := data.Get(SchemaFeature).(interface{})
 	accountFeatures := constructAccountFeatures(&cloudauth.AccountFeatures{}, featureData)
 
 	return &v2.CloudauthAccountSecure{
 		CloudAccount: cloudauth.CloudAccount{
-			Enabled:    data.Get("enabled").(bool),
-			ProviderId: data.Get("cloud_provider_id").(string),
-			Provider:   cloudauth.Provider(cloudauth.Provider_value[data.Get("cloud_provider_type").(string)]),
+			Enabled:    data.Get(SchemaEnabled).(bool),
+			ProviderId: data.Get(SchemaCloudProviderId).(string),
+			Provider:   cloudauth.Provider(cloudauth.Provider_value[data.Get(SchemaCloudProviderType).(string)]),
 			Components: accountComponents,
 			Feature:    accountFeatures,
 		},
 	}
 }
 
-func cloudauthAccountToResourceData(data *schema.ResourceData, cloudAccount *v2.CloudauthAccountSecure) error {
-	for _, err := range []error{
-		data.Set("id", cloudAccount.Id),
-		data.Set("enabled", cloudAccount.Enabled),
-		data.Set("cloud_provider_id", cloudAccount.ProviderId),
-		data.Set("cloud_provider_type", cloudAccount.Provider.String()),
-		data.Set("components", cloudAccount.Components),
-		data.Set("feature", cloudAccount.Feature),
-	} {
-		if err != nil {
-			return err
+/*
+	This helper function converts feature values from *cloudauth.AccountFeature to resource data schema.
+*/
+
+func featureValuesToResourceData(name string, feature *cloudauth.AccountFeature) map[string]interface{} {
+	valuesMap := make(map[string]interface{})
+
+	valuesMap["type"] = feature.Type.String()
+	valuesMap["enabled"] = feature.Enabled
+	valuesMap["components"] = feature.Components
+
+	featureMap := map[string]interface{}{
+		name: []map[string]interface{}{
+			valuesMap,
+		},
+	}
+
+	return featureMap
+}
+
+/*
+	This helper function converts the features data from *cloudauth.AccountFeatures to resource data schema.
+	This is needed to set the value in cloudauthAccountToResourceData().
+*/
+
+func featureToResourceData(features *cloudauth.AccountFeatures) []map[string]interface{} {
+	featureMap := []map[string]interface{}{}
+
+	featureFields := map[string]*cloudauth.AccountFeature{
+		SchemaSecureThreatDetection:     features.SecureThreatDetection,
+		SchemaSecureConfigPosture:       features.SecureConfigPosture,
+		SchemaSecureIdentityEntitlement: features.SecureIdentityEntitlement,
+		SchemaMonitorCloudMetrics:       features.MonitorCloudMetrics,
+		SchemaSecureAgentlessScanning:   features.SecureAgentlessScanning,
+	}
+
+	for name, feature := range featureFields {
+		if feature != nil {
+			value := featureValuesToResourceData(name, feature)
+			featureMap = append(featureMap, value)
 		}
 	}
+
+	return featureMap
+}
+
+func cloudauthAccountToResourceData(data *schema.ResourceData, cloudAccount *v2.CloudauthAccountSecure) error {
```

Comment:
I wish we hadn't blown the implementation of this function from the terse version we previously had.